keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -883,20 +889,51 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
-detecting pod failures. For this purpose we need to make sure that pods stuck
-in the pending phase which are terminating (with set deletionTimestamp) are
-eventually marked as failed. Such pods may either be not scheduled
-(for example when they are unschedulable due to node affinity configuration for
-which no node exists in the cluster) or scheduled (for example when they fail
-to pull the the docker image). In case of non-scheduled pods they will be moved
-to the `Failed` phase by the PodGC controller as it marks as failed all
-non-scheduled terminating pods (after the Alpha changes for this feature, see
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+A scheduled pod (i.e. a pod is bound to a node and its `spec.nodeName` indicates
+the node) may also get stuck in a `Pending` phase (for example due to an invalid
+image reference). In that case, as long as the pod is not terminating, it
+could be a transient issue and Kubelet can make progress after retrying.
+Once such a Pod (in a `Pending` phase) is also terminating, Kubelet is not going
+to attempt starting of the pod. Here, we say that a pod is terminating when its
+`deletionTimestamp` is set by API server due to a Delete request (which can be
+called on the pod by a user or a k8s component).
+
+Such cases are problematic in case of Jobs with pod failure policies, which we want
+to match only against pods in terminal phase (`Succeeded` or `Failed`), to ensure
+their state is not modified while Job controller matches them against the policy.
+
+In order to ensure that Job controller only considers pods in terminal phase we
+need to make sure pods in the following states are transitioned into `Failed`
+phase:
+1. `Running` phase, and orphaned (scheduled to a node which no longer exists)
+2. `Pending` phase, terminating, and unscheduled
+3. `Pending` phase, terminating, and scheduled
+
+For Alpha, we implemented a fix for scenarios 1. and 2. by setting the pod phase
+as `Failed` in PodGC. Note that, in both scenarios there is no Kubelet responsible
+to transition the pod, thus it is done by PodGC.
+
+As the 3rd scenario wasn't fixed in the first iteration of Beta (1.26), we only
+require pods to be in terminal phase when `podFailurePolicy` is specified (and
+only in that case), but this creates an inconsistency with handling Jobs
+with and without pod failure policies
+(see Issue: [Job controller should wait for Pods to terminate to match the failure policy](https://github.com/kubernetes/kubernetes/issues/113855)).
+
+For the second iteration of Beta (1.27), we plan to fix the 3rd scenario by
+modifying Kubelet to transition `Pending`, terminating, and scheduled pods into
+the `Failed` phase, allowing the Job
+controller to count them, match against the (optional) pod failure policy and
+remove their Job finalizers, allowing API server to complete deletion.
```

Comment:
Something worth to note here is that this will require 2x API calls from kubelet -> apiserver for these types of pods.

If before kubelet was simply deleting the pod: only one HTTP call was needed (i.e. HTTP DELETE), now this will require two HTTP calls, i.e. (1) Patch the pod status with phase (2) Delete the pod.

I assume the cases of these types of pods is relatively rare, so I don't think is a large concern, but worth to document.

Reply:
I think (please correct me if I'm missing something), that there is one additional API call needed kubelet -> API server to PATCH the pod (to move the transition phase). 

The new flow of requests in this scenario:
1. a scheduled pod (with nodeName != nil) is in Pending phase (same)
2. DELETE request to API server is sent by user or another k8s component (e.g. scheduler) -> the pod is terminating
3. [new] PATCH by Kubelet to transition the Pod into Failed phase
4. PATCH the pod to remove the finalizer by the job controller
5. Kubelet clears up the resources as there is no Pod

I'm not aware of Kubelet sending DELETE request after the PATCH (3.). Let me know if this is the case and where I can find it in code. 

Anyway, we will have one additional PATCH kubelet -> api server, but indeed the situation should be rare as it requires DELETE which needs to be send by a user or another component for the Pod to be terminating.

Reply:
Update: added a paragraph on this. PTAL.

Reply:
Kubelet will need to send the final delete request to the api-server to actually force delete the pod with zero grace period seconds (just like it does today). So today the flow is:

1. Pod created with spec.nodeName != nil
2. Scheduler sets spec.nodeName
3. Pod is pending
4. API-server delete recieves delete with some deletion grace period seconds > 0; api-server set's a deletionTimestamp on the pod
5. Kubelet sends HTTP DELETE to api-server with zero grace period seconds to remove the pod from the api-server

With this feature we need an additional call between steps 4 and 5 to send the HTTP patch to update the phase to failed.

Reply:
I see, I've checked that it is send from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L745-L758.  I guess this is to speed up the deletion in API server, because otherwise API server would wait until the original grace period elapsed?

Still it holds that there is a single additional PATCH request.

Reply:
There is a PRR question about additional API calls. Please update.

Reply:
Added a small note there.

Reply:
> I see, I've checked that it is send from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L745-L758. I guess this is to speed up the deletion in API server, because otherwise API server would wait until the original grace period elapsed?

That is the final delete sent to the API server from the kubelet to delete the pod from etcd which is what is expected in the normal case. Only the kubelet knows when it is safe to force delete the pod (since it should only force delete the pod once the pod is terminal).